### DISCUSSION TITLE
Allow for partial updates to CTS resolver, ATLAS and search index

### DISCUSSION THREAD

**Discussion by jacobwegner** [scaife-viewer/scaife-viewer.org#10] [Ideas]
Currently, any content update for SV1 or SV2 readers requires a complete re-ingestion / re-indexing of all corpora in a given site / library.

Depending on the level of annotation, work-part level annotations (like attributions or repo metadata) may be "reset" and reapplied on top of an existing library.

Some of our "bulk ingestion" improvements for ATLAS data models built on top of things like `django-treebeard` make assumptions about `path` or `idx` values (refs scaife-viewer/backend#22)...but we already have a wonderful construct in the URN that might allow us to address path-like functionality.

I've also done some experiments under [scaife-viewer/scaife-viewer#indexer-breakout](https://github.com/scaife-viewer/scaife-viewer/blob/110b790d81db9163368fb2ed989c8a29ca92bffe/scaife_viewer/indexer.py#L242) that calculates passage-level SHAs that might be useful.

@eelkevdbos and I have had some offline conversations as well about partial re-ingestion / re-indexing for a particular corpus within a larger collection of corpora via partitioning / sharding.  There would be some efficiencies for having a "central" ATLAS database, but it might be possible to load individual corpus-level databases and then use high-level SQLite (or whatever RDMS is used) functionality to "merge" datasets...the idea being that if a URN within a corpus is unique, we can use that as a primary key within the merge.

Eventually I can see creating some issues within [scaife-viewer/backend](https://github.com/scaife-viewer/backend) to tackle this.  Would likely expand some of the work done on scaife-viewer/backend#8

**Comment by jacobwegner** [comment]
This work would be crucial for establishing an "artifact pipeline" for Scaife Viewer instances:
- Curators update data sets in a source repo
- Automated pipelines run that perform quality checks, auto-generate additional annotations, etc
- Source repos trigger the artifact pipeline which would performa a "delta" update (only impacting the data directly related to the trigger)
- Once the artifact pipeline is completed, the deployment pipeline would be notified
- Deployment pipeline could use automated or manual jobs to deploy the new data to the instance

**Comment by jacobwegner** [comment]
Delta updates are becoming more critical for the builds of [scaife-viewer/beyond-translation-site](https://github.com/scaife-viewer/beyond-translation-site) and other sites running on the Scaife Viewer platform.

I'll likely do some work in that repo soon to improve the re-use of Docker layers, and may use that work as an attempt to chip away at this larger problem.

**Comment by jacobwegner** [comment]
I guess to expand on my last comment, part of this work would be splitting some of the preparation of an ATLAS artifact from the runtime of a deployment.

- In some instances, we're building and bundling ATLAS data into a Docker image.
- In others, we're doing the building within a booted container. 
- Another thing to explore would be something similar to [Kuberentes Init Containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/), e.g. [Does docker-compose support init container?](https://stackoverflow.com/questions/70322031/does-docker-compose-support-init-container) (from Stack Overflow).

(We've made use of k8s init containers in the past to help pre-populate the CTS API cache on deployments of [scaife-viewer/scaife-cts-api](https://github.com/scaife-viewer/scaife-cts-api))

**Comment by eelkevdbos** [comment]
Something that I personally would like to explore is a scenario where individual corpora are converted offline (read: ci-compatible) to their atlas sqlite representation and exported as regular SQL statements.

These SQL statements would probably need some trimming to remove any conflicting incremental identifiers and other schema inconsistencies that would otherwise lead to a "merge conflict" of sql statements.

Following the export of the individual corpora, a single combination step executes all exported sql statements in a fresh database bootstrapped with the atlas schema. Possible optimizations / application of indices can be applied after this step.

---

Some of the unknowns:

- What schema changes would be necessary to make this work?
- How does this translate into documents for indexing?
- Can we do some document 'tracing' by tagging individual documents with corpus identifiers?

**Reply by jacobwegner** [reply]
@eelkevdbos I do recall you mentioning this before 😄 

Here's just a brief response, and if we want to do something concrete for Scholarly Editions maybe we could have a more in-depth planning session.

_CTS Resolver_
The CTS resolver would be the first place we'd want to handle partial updates (or merging corpora), as SV1 sites (Scholarly Editions and scaife.perseus.org) are still pretty heavily tied to that resolver.

SV2 sites (like https://beyond-translation.perseus.org/reader/urn:cts:greekLit:tlg0012.tlg001.perseus-grc2:1.1-1.7) don't use the CTS resolver so wouldn't benefit from those optimizations.

_ATLAS SQLite db_

As far as the actual ATLAS tables, the `django-treebeard` library and the modeling we've done for ATLAS have some hard dependencies on Django `pk` fields and the `path` field from `django-treebeard` that would need to be re-evaluated.

I'd probably take a closer look at [Datasette cross-database-queries](https://docs.datasette.io/en/stable/sql_queries.html#cross-database-queries) and [sqlite-utils](https://simonwillison.net/2019/Feb/25/sqlite-utils/) to think about merging those "compiled" databases.

But even beyond "merging" those databases, there are probably still single-corpora optimizations to things in the ingestion pipeline that could be used to help do "delta" updates (only re-run a pipeline step if an input has changed, for example).

_Search_

Again, the use of the new metadata collections in Scholarly Editions differs a bit from other Scaife Viewer sites as to how things are done in the search index, but a similar "delta" updates approach within a single corpus and then merging corpora would be useful.

I mentioned in my original comment here that I've experimented with storing checksums at the passage level, and we could certainly do that higher up the CTS hierarchy too.  Something like [elasticsearch-dump](https://github.com/elasticsearch-dump/elasticsearch-dump) would make some of the "merging" aspects pretty trivial.

**Comment by jacobwegner** [comment]
The past few days, I've started on a few spikes:

- Using **elasticsearch-dump** and [sqlite-utils](https://sqlite-utils.datasette.io/en/stable/) to diff some of the content in a search index.  This is useful for the `lemma_content` generation we do under SV 1, but also can be useful elsewhere  to help query / update / delete an ES index in a performant way.
- Calculating a checksum on each CTS version when we ingest it and storing that SHA in a manifest / metadata file that exists outside of MyCapytain.

**Comment by jacobwegner** [comment]
Circling back here...been working on some things for SV 1 sites to work towards incremental deploys.

A big blocker for [scaife.perseus.org](https://scaife.perseus.org) (aka _SV-PDL_) has been the lemmatization pipeline requiring so many resources we've been spinning up enormous VMs to do two passes on indexing (from scratch); the first pass uses many parallel CPUs, the second using nearly a _terabyte_ of RAM 😱!

Work related to [Lemmatization of texts #24
](https://github.com/scaife-viewer/scaife-viewer.org/discussions/24) is underway which will decouple lemmatization from the search index and more or less knock out that second pass; that in turn will allow us to improve the automation of the _SV-PDL_ instance that will lend itself to more incremental updates.
